### DISCUSSION TITLE
When changed to TSMorphMetadataProvider - Source class for entity User not found

### DISCUSSION THREAD

**Discussion by imcm7** [mikro-orm/mikro-orm#4728] [General]
Source class for entity User not found. Verify you have 'compilerOptions.declaration' enabled in your 'tsconfig.json'. If you are using webpack, see https://bit.ly/35pPDNn

With `entities: [User],` and
```
// import * as entities from '~/server/entities;
entities: Object.values(entities),
```

https://github.com/imcm7/nuxt-db

```
Source class for entity User not found. Verify you have 'compilerOptions.declaration' enabled in your 'tsconfig.json'. If you are using webpack, see https://bit.ly/35pPDNn
  at TsMorphMetadataProvider.readTypeFromSource (./node_modules/@mikro-orm/reflection/TsMorphMetadataProvider.js:68:19)  
  at TsMorphMetadataProvider.initPropertyType (./node_modules/@mikro-orm/reflection/TsMorphMetadataProvider.js:52:41)  
  at TsMorphMetadataProvider.initProperties (./node_modules/@mikro-orm/reflection/TsMorphMetadataProvider.js:37:22)  
  at TsMorphMetadataProvider.loadEntityMetadata (./node_modules/@mikro-orm/reflection/TsMorphMetadataProvider.js:23:14)  
  at MetadataDiscovery.discoverEntity (./node_modules/@mikro-orm/core/metadata/MetadataDiscovery.js:267:31)  
  at MetadataDiscovery.discoverReferences (./node_modules/@mikro-orm/core/metadata/MetadataDiscovery.js:201:18)  
  at MetadataDiscovery.findEntities (./node_modules/@mikro-orm/core/metadata/MetadataDiscovery.js:117:14)  
  at MetadataDiscovery.discover (./node_modules/@mikro-orm/core/metadata/MetadataDiscovery.js:35:20)  
  at MikroORM.discoverEntities (./node_modules/@mikro-orm/core/MikroORM.js:151:46)  
  at MariaDbMikroORM.init (./node_modules/@mikro-orm/core/MikroORM.js:33:19)
[discovery] ORM entity discovery started, using TsMorphMetadataProvider
[discovery] - processing entity User

```

**Comment by imcm7** [comment]
I add in root tsconfig file and in server dir `compilerOptions.declaration`

**Reply by imcm7** [reply]
That config

```
```
import { User } from '~/server/entities/user.entity';

const options = defineConfig({
  // optionally you can override the base directory (defaults to `process.cwd()`)
  allowGlobalContext: false,
  baseDir: process.cwd(),
  dbName: 'nuxt',
  debug: true, // ['discovery', 'info', 'query', 'query-params'],
  dynamicImportProvider: id => import(id),
  discovery: {
    // disableDynamicFileAccess: true,
    warnWhenNoEntities: true
  },
  driver: MariaDbDriver,
  // entities: Object.values(entities),
  // entitiesTs: Object.values(entities),
  // entities: ['./server/entities'],
  // entitiesTs: ['./server/entities/*'],
  //entities: ['./server/entities/*.js'], // path to our JS entities (dist), relative to `baseDir`
  //entitiesTs: ['./server/entities/*.ts'], // path to our TS entities (source), relative to `baseDir`
   entities: [User],
  entitiesTs: [User],
  extensions: [Migrator],
  loadStrategy: LoadStrategy.JOINED,
  metadataProvider: TsMorphMetadataProvider,
  migrations: {
    tableName: 'mikro_orm_migrations', // migrations table name
    path: process.cwd() + '/migrations', // path to folder with migration files
    glob: '!(*.d).{js,ts}', // how to match migration files (all .js and .ts files, but not .d.ts)
    transactional: true, // run each migration inside transaction
    disableForeignKeys: true, // try to disable foreign_key_checks (or equivalent)
    allOrNothing: true, // run all migrations in current batch in master transaction
    emit: 'ts', // migration generation mode
  },
  password: 'changeme',
  populateAfterFlush: true,
  port: 3306,
  // registerRequestContext: false,
  // seeder: {
  //   path: './seeders',
  //   defaultSeeder: 'DatabaseSeeder',
  // },
  tsNode: true,
  user: 'root'
});
```

**Reply by B4nan** [reply]
I believe you are esbuild for compilation, which does not generate the `.d.ts` files at all (that's what `compilerOptions.declaration` is about). 

Note that this on its own won't help you with the other issue (the unknown extension), the problem is still the same (the same file gets dynamically imported).

Things are much more complicated to set up with ts-morph if you use non-standard ways of building your project (using `esbuild` is one of them, and if often means bundlings, which is another).
